esphome/components/gcja5/gcja5.cpp

```diff
@@ -0,0 +1,119 @@
+  }
+
+  if (this->available() == 0) {
+    return;
+  }
+
+  // There must now be data waiting
+  this->last_transmission_ = now;
+  uint8_t val;
+  while (this->available() != 0) {
+    this->read_byte(&val);
+    this->rx_message_.push_back(val);
+
+    // check if rx_message_ has 32 bytes of data
+    if (this->rx_message_.size() == 32) {
+      this->parse_data_();
+
+      if (have_good_data_) {
+        if (this->pm_1_0_sensor_ != nullptr)
+          this->pm_1_0_sensor_->publish_state(get_32_bit_uint_(1));
+        if (this->pm_2_5_sensor_ != nullptr)
+          this->pm_2_5_sensor_->publish_state(get_32_bit_uint_(5));
+        if (this->pm_10_0_sensor_ != nullptr)
+          this->pm_10_0_sensor_->publish_state(get_32_bit_uint_(9));
+        if (this->pmc_0_3_sensor_ != nullptr)
+          this->pmc_0_3_sensor_->publish_state(get_16_bit_uint_(13));
+        if (this->pmc_0_5_sensor_ != nullptr)
+          this->pmc_0_5_sensor_->publish_state(get_16_bit_uint_(15));
+        if (this->pmc_1_0_sensor_ != nullptr)
+          this->pmc_1_0_sensor_->publish_state(get_16_bit_uint_(17));
+        if (this->pmc_2_5_sensor_ != nullptr)
+          this->pmc_2_5_sensor_->publish_state(get_16_bit_uint_(21));
+        if (this->pmc_5_0_sensor_ != nullptr)
+          this->pmc_5_0_sensor_->publish_state(get_16_bit_uint_(23));
+        if (this->pmc_10_0_sensor_ != nullptr)
+          this->pmc_10_0_sensor_->publish_state(get_16_bit_uint_(25));
+      } else {
+        this->status_set_warning();
+        ESP_LOGV(TAG, "Have 32 bytes but not good data. Skipping.");
+      }
+
+      this->rx_message_.clear();
+    }
+  }
+}
+
+bool GCJA5Component::calculate_checksum_() {
+  uint8_t crc = 0;
+
+  for (uint8_t i = 1; i < 30; i++)
+    crc = crc ^ this->rx_message_[i];
+
+  ESP_LOGVV(TAG, "Checksum packet was (0x%02X), calculated checksum was (0x%02X)", this->rx_message_[30], crc);
+
+  return (crc == this->rx_message_[30]);
+}
+
+uint32_t GCJA5Component::get_32_bit_uint_(uint8_t start_index) {
+  return (((uint32_t) this->rx_message_[start_index + 3]) << 24) |
+         (((uint32_t) this->rx_message_[start_index + 2]) << 16) |
+         (((uint32_t) this->rx_message_[start_index + 1]) << 8) | ((uint32_t) this->rx_message_[start_index]);
+}
+
+uint16_t GCJA5Component::get_16_bit_uint_(uint8_t start_index) {
+  return (((uint32_t) this->rx_message_[start_index + 1]) << 8) | ((uint32_t) this->rx_message_[start_index]);
+}
+
+void GCJA5Component::parse_data_() {
+  ESP_LOGVV(TAG, "GCJA5 Data: ");
+  for (uint8_t i = 0; i < 32; i++) {
+    ESP_LOGVV(TAG, "  %u: 0b" BYTE_TO_BINARY_PATTERN " (0x%02X)", i + 1, BYTE_TO_BINARY(this->rx_message_[i]),
+              this->rx_message_[i]);
+  }
+
+  if (this->rx_message_[0] != 0x02 || this->rx_message_[31] != 0x03 || !this->calculate_checksum_()) {
+    ESP_LOGVV(TAG, "Discarding bad packet - failed checks.");
+    return;
+  } else
+    ESP_LOGVV(TAG, "Good packet found.");
+
+  have_good_data_ = true;
```

Comment:
Please make sure to prefix the member variables with `this->`
```suggestion
  this->have_good_data_ = true;
```